### PR TITLE
fix!: normalize Python depName in every manager

### PR DESCRIPTION
**Pull request by not7cd**
## Changes

Adds normalized `packageName` to every Python manager

## Context

Related to #16012

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/github/index.ts [hunk 1]
```diff
@@ -91,6 +91,7 @@ import type {
   PlatformConfig,
 } from './types';
 import { getAppDetails, getUserDetails, getUserEmail } from './user';
+import { normalizeDepName } from '../../datasource/pypi/common';
```
Comment (viceice) [inline lib/modules/platform/github/index.ts:94]: please move to utils folder, we don't like to have references from platform to datasource

File: lib/modules/manager/pip_requirements/extract.ts [hunk 2]
```diff
@@ -121,7 +122,8 @@ export function extractPackageFile(
       const currentValue = currVal?.trim();
       dep = {
         ...dep,
-        depName,
+        depName: normalizeDepName(depName),
+        packageName: depName,
```
Comment (viceice) [inline lib/modules/manager/pip_requirements/extract.ts:126]: shouldn't we only normalize package name? that I'd used by datasource, depName is for visualize

**Comment by Churro** [comment]
Wouldn't that break package rules (and replacement presets) that now specify the name with `-_.`? Not sure why replacing it in the pypi datasource wouldn't be enough...

File: lib/modules/manager/pip_requirements/extract.ts (hunk 2 above)
Comment (not7cd) [reply lib/modules/manager/pip_requirements/extract.ts:126]: I'm not sure. It's not documented in the interface.

Comment (not7cd) [reply lib/modules/manager/pip_requirements/extract.ts:126]: When `alertPackageRule` is generated, `depName` is used in branch name. I believe it should be normalized. Both `Pillow` and `pillow` and IIRC updates are grouped by branch names, would this mean that every valid name would generate a separate PR?

```
  "alertPackageRules": [
    {
      "matchDatasources": [
        "pypi"
      ],
      "matchPackageNames": [
        "Pillow"
      ],
      "matchCurrentVersion": "== 10.0.1",
      "matchFileNames": [
        "requirements.txt"
      ],
      "allowedVersions": "==10.2.0",
      "isVulnerabilityAlert": true,
      "force": {
        ...
        "branchTopic": "{{{datasource}}}-{{{depName}}}-vulnerability",
        "prCreation": "immediate",
        "enabled": true,
        "addLabels": [
          "security"
        ]
      }
    }
]
```

**Comment by not7cd** [comment]
Is there a way to make this change non-breaking? Would this involve some sort of normalization of the configuration itself?

**Review by rarkins — COMMENTED** [review]
What is the use case here? e.g. different upper/lower case used in the package file vs on PyPI? If so then normalizing in the datasource seems better

**Comment by not7cd** [comment]
There are several problems and they mainly comes down to comparasion logic. 
1. vulnerability alerts are case sensitive: for example `Pillow` is returned with its non-normalized name, but it may often exist in requirements in its normalized form as `pillow`. [This results in Renovate not creating PRs](https://github.com/renovatebot/renovate/discussions/27069).
2. Package name matching it may be cumbersome to match for both `package_name` and `package-name` or `Django` and `django`.

**Comment by rarkins** [comment]
https://github.com/renovatebot/renovate/issues/28546 will make matchPackageNames case-insensitive, hopefully in v38

**Comment by not7cd** [comment]
As per documentation. There are more valid names that can lead to rules not matching.

https://packaging.python.org/en/latest/specifications/name-normalization/

This means that the following names are all equivalent:

-    friendly-bard (normalized form)
-    Friendly-Bard
-    FRIENDLY-BARD
-    friendly.bard
-    friendly_bard
-    friendly--bard
-    FrIeNdLy-._.-bArD (a terrible way to write a name, but it is valid)

**Comment by rarkins** [comment]
Then we should generate appropriate matchPackagePatterns instead of matchPackageNames

**Comment by not7cd** [comment]
What do you mean by that? We as a manager? Where such logic should exist?

Still the simplest way is to compare normalized names. If I understand correctly `packageName` is such canonical name while `depName` is used fo display.

**Review by viceice — COMMENTED** [review]
@rarkins I think we can merge this with v38, as we remove the match dep name fallback anyways. so this change is acceptable 

but this PR needs tests changes and runs against real repos

**Review by rarkins — COMMENTED** [review]
Can it be called normalizePypiName() instead?

**Comment by not7cd** [comment]
I would prefer `normalizePythonPackageName` as this is general spec, as per doc linked earlier.

How this issue relates to #28551. It's desirable to include this PR with another breaking change?

Also, a workaround #28214 could be reverted/refactored with this PR.

**Comment by not7cd** [comment]
Should https://github.com/renovatebot/renovate/pull/27733/commits/7babd040765c76e4c362991193197cb652f86e51 be moved to separate PR? Maybe `normalizePythonDepName` should be moved to a different module, as it is now widely used?

**Review by rarkins — APPROVED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]
@viceice is it really breaking though?

**Comment by not7cd** [comment]
This may break configs where they match for `some_package` or `some.package`.

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs documentation explaining to people that python packageName fields are normalized

File: docs/usage/python.md [hunk 3]
```diff
@@ -23,6 +23,15 @@ Legacy versions with the `===` prefix are ignored.
 1. Renovate searches for the latest version on [PyPI](https://pypi.org/) to decide if there are upgrades
 1. If the source package includes a GitHub URL as its source, and has a "changelog" file _or_ uses GitHub releases, a Release Note will be embedded in the generated PR
 
+## Package name matching
+
+> New in v38
```
Comment (viceice) [inline docs/usage/python.md:28]: do we really need this line?

File: lib/modules/manager/pip_requirements/extract.ts [hunk 4]
```diff
@@ -122,6 +123,7 @@ export function extractPackageFile(
       dep = {
         ...dep,
         depName,
+        packageName: normalizePythonDepName(depName),
```
Comment (viceice) [inline lib/modules/manager/pip_requirements/extract.ts:126]: only set if different from depName

File: lib/modules/manager/pip_setup/extract.ts [hunk 5]
```diff
@@ -45,6 +46,7 @@ function depStringHandler(
 
   const dep: PackageDependency<ManagerData> = {
     depName,
+    packageName: normalizePythonDepName(depName),
```
Comment (viceice) [inline lib/modules/manager/pip_setup/extract.ts:49]: same

File: lib/modules/manager/setup-cfg/extract.ts [hunk 6]
```diff
@@ -73,6 +74,7 @@ function parseDep(
 
   const dep: PackageDependency = {
     depName,
+    packageName: normalizePythonDepName(depName),
```
Comment (viceice) [inline lib/modules/manager/setup-cfg/extract.ts:77]: same

File: lib/workers/repository/init/vulnerability.spec.ts [hunk 7]
```diff
@@ -367,40 +367,5 @@ describe('workers/repository/init/vulnerability', () => {
       expect(res.packageRules).toMatchSnapshot();
       expect(res.packageRules).toHaveLength(1);
     });
-
-    it('returns pip alerts with normalized name', async () => {
```
Comment (viceice) [inline lib/workers/repository/init/vulnerability.spec.ts:371]: no longer needed to validate?

I think it can still fail osv?

Comment (not7cd) [reply lib/workers/repository/init/vulnerability.spec.ts:371]: Yes, because I expect that `[github.getVulnerabilityAlerts](https://github.com/renovatebot/renovate/pull/27733/commits/a1fbc42083aa469af60b78ddca8ff66f64355972)` returns normalized names. So the mock here doesn't make sense. Maybe an error could be thrown in `workers/repository/init/vulnerability`?

File: docs/usage/python.md (hunk 3 above)
Comment (rarkins) [reply docs/usage/python.md:28]: No

File: lib/modules/manager/pep621/utils.ts [hunk 8]
```diff
@@ -60,7 +61,7 @@ export function pep508ToPackageDependency(
   }
 
   const dep: PackageDependency = {
-    packageName: parsed.packageName,
+    packageName: normalizePythonDepName(parsed.packageName),
```
Comment (not7cd) [inline lib/modules/manager/pep621/utils.ts:64]: pep621 always sets `packageName`, even before this change.

File: lib/modules/manager/poetry/schema.ts [hunk 9]
```diff
@@ -158,7 +158,7 @@ export const PoetryDependencies = LooseRecord(
   for (const [depName, dep] of Object.entries(record)) {
     dep.depName = depName;
     if (!dep.packageName) {
-      const packageName = normalizeDepName(depName);
+      const packageName = normalizePythonDepName(depName);
```
Comment (not7cd) [inline lib/modules/manager/poetry/schema.ts:161]: Poetry manager also always sets `packageName`

File: lib/modules/manager/pip-compile/extract.ts [hunk 10]
```diff
@@ -257,18 +254,18 @@ function extendWithIndirectDeps(
...
     }
   }
 }
 
+function compareNames(a: PackageDependency, b: PackageDependency): boolean {
+  const aName = a.packageName ?? a.depName;
+  const bName = b.packageName ?? b.depName;
+  return aName === bName;
+}
+
```
Comment (not7cd) [inline lib/modules/manager/pip-compile/extract.ts:268]: I don't like this, is there a more reliable way to compare names? Doing it inline didn't work.

**Review by not7cd — COMMENTED** [review]
I don't really want to agree that conditional logic added in https://github.com/renovatebot/renovate/pull/27733/commits/bfa8fa9cd7cd0d8f680a156be0399fb0d502a16b makes the code better. This forced me to create this comparison function that does a lot of `??` magic, which I think I understand. 

This also doesn't match the current definition in the interface itself. I know that I have created this doc entry, but it's based on #16012.
https://github.com/renovatebot/renovate/blob/01d4d0db8af4037d619e04c7ae21848acf9e6682/lib/modules/manager/types.ts#L104-L107

This is also inconsistent with other managers like pep621 and poetry.

If there is a consensus and if needed https://github.com/renovatebot/renovate/pull/27733/commits/bfa8fa9cd7cd0d8f680a156be0399fb0d502a16b can be reverted without my interaction, in case I am unresponsive. I'm not sure when do you want to release v38.

**Comment by jeroenmaas** [comment]
We're running into similar (or same) issue. https://github.com/renovatebot/renovate/discussions/29230 Great work on this [not7cd](https://github.com/not7cd). Hope this could be resolved soon.

**Comment by not7cd** [comment]
[My objections](https://github.com/renovatebot/renovate/pull/27733#pullrequestreview-2018268937) to changes related to [this change request](https://github.com/renovatebot/renovate/pull/27733#discussion_r1574967662) are still unresolved. I'm pointing this out as they are now hidden in this thread.

File: lib/workers/repository/init/vulnerability.ts [hunk 11]
```diff
@@ -200,11 +200,8 @@ export async function detectVulnerabilityAlerts(
           matchFileNames,
         };
 
-        if (
-          datasource === PypiDatasource.id &&
-          normalizePythonDepName(depName) !== depName
-        ) {
-          matchRule.matchPackageNames?.push(normalizePythonDepName(depName));
+        if (datasource === PypiDatasource.id) {
+          matchRule.matchPackageNames = [normalizePythonDepName(depName)];
```
Comment (not7cd) [inline lib/workers/repository/init/vulnerability.ts:204]: I think this it should be removed, but I don't remember at this point. And I could miss it during cherry-pick.

**Review by rarkins — APPROVED** [review]

File: lib/modules/manager/setup-cfg/extract.ts [hunk 12]
```diff
@@ -73,11 +74,11 @@ function parseDep(
...
     currentValue,
     datasource: PypiDatasource.id,
     depType,
   };
-
```
Comment (viceice) [inline lib/modules/manager/setup-cfg/extract.ts:80]: revert?

File: lib/modules/platform/github/index.ts [hunk 13]
```diff
@@ -36,8 +36,8 @@ import type {
 } from '../../../util/git/types';
 import * as hostRules from '../../../util/host-rules';
 import { repoCacheProvider } from '../../../util/http/cache/repository-http-cache-provider';
-import * as githubHttp from '../../../util/http/github';
 import type { GithubHttpOptions } from '../../../util/http/github';
+import * as githubHttp from '../../../util/http/github';
```
Comment (viceice) [inline lib/modules/platform/github/index.ts:40]: 🤔 why this change?

**Review by viceice — COMMENTED** [review]
just two small nits 🙃

File: lib/modules/platform/github/index.ts (hunk 13 above)
Comment (not7cd) [reply lib/modules/platform/github/index.ts:40]: from VS Code organize imports

**Review by viceice — APPROVED** [review]
